### PR TITLE
Generate docs in JSON format

### PR DESCRIPTION
**Pull request by david-crespo**
The existing docs site is powered by markdown files generated from here, which means the display logic is essentially encoded into the CLI rather than the site. This PR produces a single JSON file instead, which lets us pull it with a single request and leaves the display logic up to the new docs site. It is TBD how we will actually get this into the docs site. The simplest thing would probably be to check the file in so we can just make a request for a particular version at build time like we're doing with the Nexus OpenAPI spec.

Note that top-level metadata is the same as all the lower-level metadata, i.e., there's no "app name" or whatever, `oxide` is just a command like all the other ones, with `excerpt`, `about` etc. Simple.

### TypeScript types for the doc

We'll use these on the docs site side to represent this data. I've made it so if fields are going to be missing, we leave them out rather than leaving the key in with a value of `null`. This can be annoying when you don't have types, but I think it's fine here, and it takes a lot of noise out of the output.

We can also use a Zod schema on the docs side when we pull this JSON (probably at build time) to validate that it matches this type.

```ts
// if there is nothing in here we filter it out of the list
type Arg = {
  short?: string
  long?: string
  help?: string
}

type CliCmd = {
  title: string
  excerpt: String
  about?: string
  args?: Arg[]
  subcommands?: CliCmd[]
}
```

### Output

```json
{
  "title": "oxide",
  "excerpt": "Work seamlessly with Oxide from the command line",
  "about": "Work seamlessly with Oxide from the command line.\n\nEnvironment variables that can be used with oxide. Additionally to those listed below, some flags have a corresponding environment variable. For example, most of the time, the `--organization,-o` flag is mapped to the `OXIDE_ORG` environment variable.\n\nOXIDE_TOKEN: an authentication token for Oxide API requests. Setting this avoids being prompted to authenticate and takes precedence over previously stored credentials.\n\nOXIDE_HOST: specify the Oxide hostname for commands that would otherwise assume the \"api.oxide.computer\" host.\n\nOXIDE_BROWSER, BROWSER (in order of precedence): the web browser to use for opening links.\n\nDEBUG: set to any value to enable verbose output to standard error.\n\nOXIDE_PAGER, PAGER (in order of precedence): a terminal paging program to send standard output to, e.g. \"less\".\n\nNO_COLOR: set to any value to avoid printing ANSI escape sequences for color output.\n\nCLICOLOR: set to \"0\" to disable printing ANSI colors in output.\n\nCLICOLOR_FORCE: set to a value other than \"0\" to keep ANSI colors in output even when the output is piped.\n\nOXIDE_FORCE_TTY: set to any value to force terminal-style output even when the output is redirected. When the value is a number, it is interpreted as the number of columns available in the viewport. When the value is a percentage, it will be applied against the number of columns available in the current viewport.\n\nOXIDE_NO_UPDATE_NOTIFIER: set to any value to disable update notifications. By default, oxide checks for new releases once every 24 hours and displays an upgrade notice on standard error if a newer version was found.\n\nOXIDE_CONFIG_DIR: the directory where oxide will store configuration files. Default: \"$XDG_CONFIG_HOME/oxide\" or \"$HOME/.config/oxide\".\n\nAuthentication\n\nYou can get an access token running `oxide auth login`. This will contact `OXIDE_HOST` and attempt an OAuth 2.0 Device Authorization Grant. The CLI will attempt to open a browser window with which you can login (via SAML or other IdP method) and type in or verify the user code printed in the terminal. After a successful login and code verification, a token associated with the logged-in user will be granted and stored in the config file.",
  "args": [
    {
      "short": "h",
      "long": "help",
      "help": "Print help information"
    },
    {
      "short": "V",
      "long": "version",
      "help": "Print version information"
    },
    {
      "short": "d",
      "long": "debug",
      "help": "Print debug info"
    }
  ],
  "subcommands": [
    {
      "title": "alias",
      "excerpt": "Create command shortcuts.",
      "about": "Create command shortcuts.\n\nAliases can be used to make shortcuts for oxide commands or to compose multiple commands.\nRun \"oxide help alias set\" to learn more.",
      "args": [
        {
          "short": "h",
          "long": "help",
          "help": "Print help information"
        },
        {
          "short": "d",
          "long": "debug",
          "help": "Print debug info"
        }
      ],
```

```json 
{
  "title": "instance",
  "excerpt": "Create, list, edit, view, and delete instances.",
  "about": "Create, list, edit, view, and delete instances.\n\nAdditionally, start, stop, and reboot instances.",
  "args": [
    {
      "short": "h",
      "long": "help",
      "help": "Print help information"
    },
    {
      "short": "d",
      "long": "debug",
      "help": "Print debug info"
    }
  ],
  "subcommands": [
    {
      "title": "disks",
      "excerpt": "List the disks attached to an instance.",
      "args": [
        {
          "short": "p",
          "long": "project",
          "help": "The project that holds the instance"
        },
        {
          "short": "o",
          "long": "organization",
          "help": "The organization to view the project"
        },
        {
          "short": "f",
          "long": "format",
          "help": "Output format"
        },
        {
          "short": "h",
          "long": "help",
          "help": "Print help information"
        },
        {
          "short": "d",
          "long": "debug",
          "help": "Print debug info"
        }
      ]
    },
```

### CODE REVIEW

**Comment by david-crespo** [comment]
Requesting a quick look from @ahl as our resident JSON spec generation expert.

File: src/cmd_generate.rs [hunk 1]
```diff
@@ -203,6 +294,118 @@ mod test {
...
+            debug: false,
+        };
+
+        let cmd = crate::cmd_generate::CmdGenerateJson { dir: "".to_string() };
+        let app = crate::cmd_generate::test_app();
+
+        let json = cmd.generate(&mut ctx, &app).unwrap();
+
+        assert_eq!(
+            serde_json::to_string_pretty(&json).unwrap(),
```
Comment (david-crespo) [inline src/cmd_generate.rs]: this is a bit fake but it's a lot easier than calling the serialize to file business and asserting about the file. or is it?

**Review by karencfv — COMMENTED** [review]
Looks good to me! Eventually we'll want to have examples as well, but we can add that when we actually have the examples.

I think for now we should generate the JSON file and commit it to this repo in a `docs/` directory.

File: src/cmd_generate.rs (hunk 1 above)
Comment (karencfv) [reply src/cmd_generate.rs]: This is enough I think

**Comment by david-crespo** [comment]
We could use `expectorate` like we do in Omicron and avoid the whole bot commit thing. I’ll give that a shot. It also solves my concern with that test being fake.

**Comment by karencfv** [comment]
How about using the CLI itself to generate the JSON with a Make target? https://github.com/oxidecomputer/cli/blob/main/Makefile#L142-L151

That's a good way of testing too!

**Comment by karencfv** [comment]
We can then hook that make target to the release action to make sure it's up to date :)

File: src/cmd_generate.rs [hunk 2]
```diff
@@ -199,10 +290,23 @@ fn test_app() -> clap::Command<'static> {
...
+    #[test]
+    fn test_generate_json() {
+        let output = Exec::shell("cargo run -- generate json")
+            .stdout(Redirection::Pipe)
+            .capture()
+            .unwrap()
+            .stdout_str();
+        assert_contents("docs/oxide.json", &output);
+    }
+
```
Comment (david-crespo) [inline src/cmd_generate.rs:309]: @karencfv I based this on the way we keep the OpenAPI spec up to date in Omicron — this test fails if a change is made but `docs/oxide.json` isn't updated manually, whether by running `cargo run -- generate json -D docs` or by running this test with `EXPECTORATE=overwrite`.

https://github.com/oxidecomputer/omicron/blob/88dd1e1606e611ab30b976c57da5ef1859d40a44/nexus/tests/integration_tests/commands.rs#L104-L168

I'm just using `cargo run` here — do you see an advantage to using a make command instead? Or even something else?

Comment (karencfv) [reply src/cmd_generate.rs:309]: I guess the benefit I saw is that it's the way it's currently done, but if we're going to get rid of that GH action anyway, it shouldn't make a difference I suppose 🤔 

Could you add a section on the README on how the docs are generated?

**Comment by david-crespo** [comment]
Pretty baffling test failure considering all I changed is the README – sure seems in the below screenshot like something is carrying over from #246, which removed `pager` and was just merged into main. Is it running the tests for `main` instead of this commit?

image

**Comment by karencfv** [comment]
You're going to kill me, but it looks like you need to rebase/merge from the main branch and run the doc generation again to get the tests passing 😄

**Comment by karencfv** [comment]
It's because I removed a command :P

**Comment by david-crespo** [comment]
Yeah, but why would this care about what's on main?

**Comment by karencfv** [comment]
Sorry, looks like we wrote at the same time, didn't read your comment. I'm guessing it's testing against main? 🤷‍♀️

**Review by karencfv — APPROVED** [review]
Thanks for adding docs about the docs! 🙇‍♀️

**Comment by karencfv** [comment]
Hmmm, that didn't work... I think it's probably best to merge/rebase from main. Sorry about the pain from my previous commit

**Comment by david-crespo** [comment]
Very weird.

**Comment by karencfv** [comment]
🎉

**Pull request merged**